src/ndt7-upload-worker.js

```diff
@@ -0,0 +1,161 @@
+if (typeof WebSocket === 'undefined') {
+  global.WebSocket = require('isomorphic-ws');
+}
+
+// WebWorker that runs the ndt7 upload test
+const workerMain = function(ev) {
+  // TODO: Figure out the right place to choose wss vs ws
+  // let url = new URL(config.baseURL)
+  // url.protocol = (url.protocol === "https:") ? "wss:" : "ws:"
+  // url.pathname = "/ndt/v7/upload"
+  const url = ev.data['ws:///ndt/v7/upload'];
+  const sock = new WebSocket(url, 'net.measurementlab.ndt.v7');
+  let now = () => new Date().getTime();
+  if (typeof performance !== 'undefined' &&
+      typeof performance.now !== 'undefined') {
+    now = performance.now;
+  }
+  uploadTest(sock, postMessage, now);
+};
+
+const uploadTest = function(sock, postMessage, now) {
+  let closed = false;
+  sock.onclose = function() {
+    if (!closed) {
+      closed = true;
+      postMessage({
+        MsgType: 'complete',
+      });
+    }
+  };
+
+  sock.onmessage = function(ev) {
+    if (typeof ev.data !== 'undefined') {
+      postMessage({
+        MsgType: 'measurement',
+        Source: 'server',
+        ServerMessage: ev.data,
+      });
+    }
+  };
+
+  /**
+   * uploader is the main loop that uploads data in the web browser. It must
+   * carefully balance a bunch of factors:
+   *   1) message size determines measurement granularity on the client side,
+   *   2) the JS event loop can only fire off so many times per second, and
+   *   3) websocket buffer tracking seems inconsistent between browsers.
+   *
+   * Because of (1), we need to have small messages on slow connections, or
+   * else this will not accurately measure slow connections. Because of (2), if
+   * we use small messages on fast connections, then we will not fill the link.
+   * Because of (3), we can't depend on the websocket buffer to "fill up" in a
+   * reasonable amount of time.
+   *
+   * So on fast connections we need a big message size (one the message has
+   * been handed off to the browser, it runs on the browser's fast compiled
+   * internals) and on slow connections we need a small message. Because this
+   * is used as a speed test, we don't know before the test which strategy we
+   * will be using, because we don't know the speed before we test it.
+   * Therefore, we use a strategy where we grow the message exponentially over
+   * time and maintain the invariant that the message size is always either 8k
+   * or less than 1/8 of the total number of bytes we have enqueued. In an
+   * effort to be kind to the memory allocator, we always double the message
+   * size instead of growing it by e.g. 1.3x.
+   *
+   * @param {*} data
+   * @param {*} start
+   * @param {*} end
+   * @param {*} previous
+   * @param {*} total
+   */
+  function uploader(data, start, end, previous, total) {
+    if (closed) {
+      // socket.send() with too much buffering causes socket.close(). We only
+      // observed this behaviour with pre-Chromium Edge.
```

Comment:
Are you noticing this in safari as well? I have ran the code locally and see this happening in Safari Version 13.1.1 (15609.2.9.1.2).

Reply:
Does https://bassosimone.github.io/ndt7/ work for you with Safari? I just checked and it seems to work for me when using Safari from mobile and from Desktop, but I only have 20 Mbit/s in upload. If you confirm that it works, then maybe we can see what are the differences in the two implementation and apply fixes. (All implementations of ndt7 in JavaScript that I know of are derived from the same small client, but there are some differences now.)

Reply:
The test in that link seems to work without any socket errors, though it seems to only send part of the upload test, it seems the test is timing out early, the max packets i am seeing is 2.1MB in size

Reply:
What does the bottom of the console looks like? Mine is like this for a test run using Safari 13.1.2:

```
[Log] {AppInfo: {ElapsedTime: 8643000, NumBytes: 20257344}, Origin: "client", Test: "upload"} (ndt7, line 221)
[Log] {AppInfo: {ElapsedTime: 8904000, NumBytes: 21305881}, Origin: "client", Test: "upload"} (ndt7, line 221)
[Log] {AppInfo: {ElapsedTime: 9156000, NumBytes: 21305856}, Origin: "client", Test: "upload"} (ndt7, line 221)
[Log] {AppInfo: {ElapsedTime: 9407000, NumBytes: 21305856}, Origin: "client", Test: "upload"} (ndt7, line 221)
[Log] ndt test complete – {Origin: "client", Test: "upload", WorkerInfo: {ElapsedTime: 10003000, Error: "Terminated with timeout"}} (ndt7, line 216)
[Log] ndt: complete (ndt7, line 123)
```

Thank you for checking!

Reply:
I work on the same team as Matt and have faster internet (300 mbps upload) and this is what I'm seeing on Safari 13.1.2:

<img width="1672" alt="Screen Shot 2020-08-17 at 2 01 26 PM" src="https://user-images.githubusercontent.com/875907/90434303-ce58e200-e092-11ea-9390-89997ac27aea.png">

Here is the tail of the console as well:

![image](https://user-images.githubusercontent.com/875907/90434417-f9dbcc80-e092-11ea-913c-3d1dac141fc3.png)

Reply:
Thanks a bunch! That is very useful to know! Maybe I can try using Docker and localhost to see if I can reproduce!